yjit_codegen.c

```diff
@@ -563,102 +570,223 @@ guard_self_is_object(codeblock_t *cb, x86opnd_t self_opnd, uint8_t *side_exit, c
-    x86opnd_t ivar_opnd = mem_opnd(64, REG0, sizeof(VALUE) * ivar_index);
-    mov(cb, REG0, ivar_opnd);
+            // Load the variable
+            x86opnd_t ivar_opnd = mem_opnd(64, REG0, offsetof(struct RObject, as.ary) + ivar_index * SIZEOF_VALUE);
+            mov(cb, REG1, ivar_opnd);
 
-    // Check that the ivar is not Qundef
-    cmp(cb, REG0, imm_opnd(Qundef));
-    je_ptr(cb, side_exit);
+            // Guard that the variable is not Qundef
+            cmp(cb, REG1, imm_opnd(Qundef));
+            je_ptr(cb, COUNTED_EXIT(side_exit, getivar_undef));
 
-    // Push the ivar on the stack
-    x86opnd_t out_opnd = ctx_stack_push(ctx, T_NONE);
-    mov(cb, out_opnd, REG0);
+            // Push the ivar on the stack
+            x86opnd_t out_opnd = ctx_stack_push(ctx, T_NONE);
+            mov(cb, out_opnd, REG1);
+        }
+        else {
+            // Compile time self is *not* embeded.
+
+            // Guard that self is *not* embedded
+            // See ROBJECT_IVPTR() from include/ruby/internal/core/robject.h
+            x86opnd_t flags_opnd = member_opnd(REG0, struct RBasic, flags);
+            test(cb, flags_opnd, imm_opnd(ROBJECT_EMBED));
+            jit_recompiling_guard(jcc_jnz, jit, ctx, getivar_max_depth, side_exit);
+
+            // check that the extended table is big enough
+            if (ivar_index >= ROBJECT_EMBED_LEN_MAX + 1) {
+                // Check that the slot is inside the extended table (num_slots > index)
+                x86opnd_t num_slots = mem_opnd(32, REG0, offsetof(struct RObject, as.heap.numiv));
+                cmp(cb, num_slots, imm_opnd(ivar_index));
+                jle_ptr(cb, COUNTED_EXIT(side_exit, getivar_idx_out_of_range));
+            }
+
+            // Get a pointer to the extended table
+            x86opnd_t tbl_opnd = mem_opnd(64, REG0, offsetof(struct RObject, as.heap.ivptr));
+            mov(cb, REG0, tbl_opnd);
+
+            // Read the ivar from the extended table
+            x86opnd_t ivar_opnd = mem_opnd(64, REG0, sizeof(VALUE) * ivar_index);
+            mov(cb, REG0, ivar_opnd);
+
+            // Check that the ivar is not Qundef
+            cmp(cb, REG0, imm_opnd(Qundef));
+            je_ptr(cb, COUNTED_EXIT(side_exit, getivar_undef));
+
+            // Push the ivar on the stack
+            x86opnd_t out_opnd = ctx_stack_push(ctx, T_NONE);
+            mov(cb, out_opnd, REG0);
+        }
 
-    return YJIT_KEEP_COMPILING;
+
+        // Jump to next instruction. This allows guard chains to share the same successor.
+        {
+            ctx_t reset_depth = *ctx;
+            reset_depth.chain_depth = 0;
+
+            blockid_t jump_block = { jit->iseq, jit_next_insn_idx(jit) };
+
+            // Generate the jump instruction
+            gen_direct_jump(
+                &reset_depth,
+                jump_block
+            );
+        }
+
+        return YJIT_END_BLOCK;
+    }
+
+    jmp_ptr(cb, side_exit);
+    return YJIT_END_BLOCK;
```

Comment:
Why not just return `YJIT_CANT_COMPILE` here instead of jumping to a side_exit?

Reply:
Because `CANT_COMPILE` emits an exit that dispatches through `pc`. It forms can form a loop with chain depth zero if `pc` is an YJIT entry point.

Reply:
Aaaaah ok cool. Let's add a comment to explain that and then I think we are ready to merge.